### PR TITLE
SASS-3552 API#1897 (for tax year 23-24) and API#16 (API#1426) Integra…

### PR DESCRIPTION
**Pull request by kannanramakrishnan**
…tion - Tax year specific API changes

### Description
Please include a brief summary of the change you've made and the issue that it fixes.

Add a link to the relevant story in Jira
[SASS-CHANGE-ME](ADD URL HERE)

##### Before Reviewing
- [ ]  Have you pulled the branch down?
- [ ]  Have you assigned yourself to the PR?
- [ ]  Have you moved the task to “in review” on JIRA?
- [ ]  Have you checked to ensure all dependencies are up to date?
- [ ]  Have you checked to ensure its been rebased against the current version of main?

##### Whilst Reviewing
- [ ]  Have you run the tests?
- [ ]  Have you run the journey tests?
- [ ]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [ ]  Have you checked for merge conflicts or any changes in the current main that may affect the current pull request? i.e. does it need another rebase?
- [ ]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [ ]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [ x]  Have you run the tests?
- [ ]  Have you run the journey tests? (where applicable)
- [ ]  Have you addressed warnings where appropriate?
- [ ]  Have you rebased against the current version of main?
- [ x]  Have you checked code coverage isn’t lower than previously?

##### After PRs been raised
- [ x]  Have you checked the PR Builder passes?

### CODE REVIEW

File: app/connectors/httpParsers/APIParser.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+      (apiError, apiErrors) match {
+        case (Some(apiError), _) => Left(ErrorModel(status, apiError))
+        case (_, Some(apiErrors)) => Left(ErrorModel(status, apiErrors))
+        case _ =>
+          pagerDutyLog(UNEXPECTED_RESPONSE_FROM_API, Some(s"[$parserName][read] Unexpected Json response."))
+          Left(ErrorModel(status, ErrorBodyModel.parsingError))
+      }
+    } catch {
+      case _: Exception => Left(ErrorModel(status, ErrorBodyModel.parsingError))
```
Comment (avelx) [inline app/connectors/httpParsers/APIParser.scala:57]: I would think that we need to log error

File: app/models/ErrorBody.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    }
+  }
+}
+
+/** Single Error **/
+case class ErrorBodyModel(code: String, reason: String) extends ErrorBody
+
+object ErrorBodyModel {
+  implicit val formats: OFormat[ErrorBodyModel] = Json.format[ErrorBodyModel]
+  val parsingError: ErrorBodyModel = ErrorBodyModel("PARSING_ERROR", "Error parsing response from API")
```
Comment (avelx) [inline app/models/ErrorBody.scala:37]: we might need to have exact API number which fail in order to simplify investigaiton

File: app/services/LiabilityCalculationService.scala [hunk 3]
```diff
@@ -16,17 +16,24 @@
...
 
-class LiabilityCalculationService @Inject()(liabilityCalculationConnector: LiabilityCalculationConnector) {
+class LiabilityCalculationService @Inject()(liabilityCalculationConnector: LiabilityCalculationConnector,
+                                            postCalculateIncomeTaxLiabilityConnector: PostCalculateIncomeTaxLiabilityConnector) {
 
-  def calculateLiability(nino: String, taxYear: String, crystallise: Boolean)(implicit hc: HeaderCarrier): Future[LiabilityCalculationResponse] = {
-    liabilityCalculationConnector.calculateLiability(nino, taxYear, crystallise)
+  def calculateLiability(nino: String, taxYear: String, crystallise: Boolean)
+                        (implicit hc: HeaderCarrier): Future[LiabilityCalculationResponse] = {
+    if (taxYear.equals("2024")) {
```
Comment (avelx) [inline app/services/LiabilityCalculationService.scala:31]: we are planning to use another API for the taxYear > 2024, so it seem that pattern matching would be intuitive choice here as more changes is expected

File: app/connectors/IFConnector.scala [hunk 4]
```diff
@@ -29,14 +29,15 @@ trait IFConnector {
   val appConfig: AppConfig
 
   val headerCarrierConfig: Config = HeaderCarrier.Config.fromConfig(ConfigFactory.load())
+  val PostCalculateIncomeTaxLiability = "1897"
```
Comment (avelx) [inline app/connectors/IFConnector.scala:32]: this is not the right place to have API specific reference under the trait which is a for a general use

File: app/config/AppConfig.scala [hunk 5]
```diff
@@ -58,6 +58,7 @@ class BackendAppConfig @Inject()(config: Configuration, servicesConfig: Services
   val environment: String = config.get[String]("microservice.services.des.environment")
   val authorisationToken: String = config.get[String]("microservice.services.des.authorisation-token")
   def iFAuthorisationToken(api:String): String = config.get[String](s"microservice.services.if.authorisation-token.$api")
+  val ifEnvironment: String = servicesConfig.getString(key = "microservice.services.if.environment")
```
Comment (avelx) [inline app/config/AppConfig.scala:61]: pls apply corresponding change in the line 58 to have distinct environment refs for IF and DES

**Review by juamalBlackman — APPROVED** [review]

**Pull request merged**